src/or/hs_descriptor.c

```diff
@@ -1583,18 +1677,30 @@ desc_decrypt_encrypted(const hs_descriptor_t *desc,
 {
   size_t encrypted_len = 0;
   char *encrypted_plaintext = NULL;
+  uint8_t *descriptor_cookie = NULL;
 
   tor_assert(desc);
+  tor_assert(desc->superencrypted_data.clients);
   tor_assert(decrypted_out);
 
-  /* XXX: We need to decrypt the descriptor properly when the client auth
-   * is enabled. */
-  (void) client_sk;
+  /* If the client secret key is provided, try to find a valid descriptor
+   * cookie. Otherwise, leave it NULL. */
+  if (client_sk) {
```

Comment:
Hmm, this "We have desc auth enabled, find descriptor cookie" block would be nice if it had its own function. Perhaps something like `find_client_auth_desc_cookie()`.

Also, I wonder if we should throw a special `log_warn()` if we have client auth enabled and we couldn't find the descriptor cookie.